tests/test-screenshot.lua

```diff
@@ -478,4 +479,38 @@ table.insert(steps, function()
     return true
 end)
 
+table.insert(steps, function()
+    local ss = awful.screenshot { auto_save_delay = 10 }
+
+    -- Reach into some `if`.
+    assert(not ss.selected_geometry)
+    assert(not ss.surface)
+    assert(#ss.surfaces == 0)
+    ss.auto_save_delay = 0
+    ss.minimum_size = 2
+    ss.minimum_size = {width = 1, height = 1}
+    ss.minimum_size = nil
+    assert(ss.surface)
+    assert(next(ss.surfaces) ~= nil)
+
+    local count = 0
+
+    local err = gdebug.print_error
+
+    function gdebug.print_error()
+        count = count + 1
+    end
+
+    -- Cause some validation failures.
+    ss.prefix = "//////"
+    assert(count == 1)
+    ss.directory = "/tmp/////"
+    ss.directory = "/tmp"
+    ss.directory = "/root/"
+    assert(count == 2)
+
+    gdebug.print_error = err
```

Comment:
Perhaps, but the reason why busted is an optional dependency is this:

```sh
> docker run --rm -ti --net=host fedora:latest /bin/bash
[root@lepagee-thinkpad-ubuntu /]# dnf search busted
Fedora 36 - x86_64                                                                                  1.2 MB/s |  81 MB     01:09    
Fedora 36 openh264 (From Cisco) - x86_64                                                            1.8 kB/s | 2.5 kB     00:01    
Fedora Modular 36 - x86_64                                                                          732 kB/s | 2.4 MB     00:03    
Fedora 36 - x86_64 - Updates                                                                        1.3 MB/s |  27 MB     00:21    
Fedora Modular 36 - x86_64 - Updates                                                                883 kB/s | 2.8 MB     00:03    
No matches found.
[root@lepagee-thinkpad-ubuntu /]# dnf search awesome
Last metadata expiration check: 0:00:44 ago on Tue Sep  6 21:39:04 2022.
================================================== Name Exactly Matched: awesome ===================================================
awesome.x86_64 : Highly configurable, framework window manager for X. Fast, light and extensible
```

Adding new dependencies in features we want contributors to use (like, writing tests) adds friction. Then *when* we make a release, it means it gets ignored upstream because the bots cannot do the upgrades. `luarock` is not really a good option. It's forbidden for distro packages and it often use a different version of Lua than the one used to package Awesome, which means it plain doesn't work for these users. Could we get better assert messages and more elegant testing than the `assert` used in the integration and doc examples? Sure, but adding *even more* friction for contributions is IMHO way worst. An alternative might be to provide our own SDK in a docker container. But that doesn't solve the distribution upgrade side of the issue. 

Plus, in the specific case, we have some code to catch all calls to `print_error` and make the test fail, so it has to be neutered if we test the error code paths.

Reply:
> Adding new dependencies in features we want contributors to use (like, writing tests) adds friction.

Doesn't that apply to docs, too? Because we just merged more than a dozen new, custom LDoc directives and a system that yells at the contributor that doesn't use them. And I'm not aware of any documentation on what directives exist or how to use them.
Whereas this scenario shows that stubs are nice-to-have, but in no way mandatory for a test to pass. Nothing would yell at a contributor for not using them. They could write their entire test suite without it, and it would be good enough.
Though, honestly, a contributor that can't copy&paste the stub example from busted's documentation will need significantly more guidance on the integration test framework as a whole. And if using stubs is friction, then that already exists in the spec tests, where we already use luassert.

But maybe I'm just missing the pattern for what's acceptable friction and what isn't.

> Then when we make a release, it means it gets ignored upstream because the bots cannot do the upgrades. 

As far as I can tell, Fedora wouldn't care. They don't run any of the test pipeline, even the parts that could be run without any dependencies.